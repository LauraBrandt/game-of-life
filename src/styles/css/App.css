.app {
  min-height: 100vh;
  position: relative;
  background-color: #900C3F; }

.app-header {
  text-align: center;
  font-family: 'Nova Mono', monospace;
  font-size: 1.3rem;
  color: #fafafa;
  text-shadow: 1px 1px 3px rgba(0, 0, 0, 0.4); }
  .app-header h1 {
    margin: 0;
    margin-right: 348px;
    padding: 0.3em 0.5em 0.4em 0.5em; }
    @media screen and (max-width: 1000px) {
      .app-header h1 {
        margin: 0; } }

.app-main {
  display: flex;
  justify-content: center;
  align-items: flex-start;
  padding-bottom: 90px; }
  @media screen and (max-width: 1000px) {
    .app-main {
      flex-direction: column;
      align-items: center; } }

.app-footer {
  position: absolute;
  bottom: 0;
  width: 100%;
  height: 60px;
  text-align: center;
  display: flex;
  justify-content: center;
  align-items: center;
  color: #e1e1e1; }
  .app-footer a {
    color: #fafafa; }
