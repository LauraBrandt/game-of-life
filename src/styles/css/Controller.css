.controller .controller-controls, .controller .controller-start-config {
  background-color: #ffc30f;
  border-radius: 8px;
  margin: 1rem;
  padding: 0.5em 0; }
  .controller .controller-controls p, .controller .controller-start-config p {
    padding: 0 0.5em 0.4em 0.5em;
    margin: 0; }

.controller {
  font-family: "Montserrat", sans-serif;
  display: flex;
  margin: 0 1em; }
  @media screen and (max-width: 770px) {
    .controller {
      font-size: 0.9em;
      flex-direction: column; } }
  .controller .controller-controls {
    flex: 4;
    height: max-content; }
    .controller .controller-controls .arrows {
      display: flex;
      justify-content: space-around;
      align-items: center; }
      .controller .controller-controls .arrows p {
        margin: 0;
        padding: 0;
        flex: 3;
        text-align: center; }
      .controller .controller-controls .arrows .arrow {
        font-size: 1.3rem;
        color: white;
        padding: 0.2rem;
        width: 2.5rem;
        height: 2.5rem;
        margin: 0.3rem;
        font-family: "Montserrat", sans-serif;
        border: 1px solid #aaa;
        border-radius: 5px;
        background-color: #e5ac00;
        cursor: pointer;
        flex: 1;
        align-self: stretch; }
        .controller .controller-controls .arrows .arrow:hover, .controller .controller-controls .arrows .arrow:focus {
          outline: none;
          background-color: #efb400; }
        .controller .controller-controls .arrows .arrow:active {
          font-size: 1.2rem;
          background-color: #fabb00; }
        @media screen and (max-width: 770px) {
          .controller .controller-controls .arrows .arrow {
            font-size: 0.9em; } }
  .controller .controller-start-config {
    flex: 5; }
    .controller .controller-start-config span {
      color: rgba(0, 0, 0, 0.7); }
  .controller .controller-button {
    display: block;
    width: 100%;
    color: white;
    border: none;
    padding: 0.4em;
    height: 2.1rem;
    font-size: 1em;
    font-family: "Montserrat", sans-serif;
    border-bottom: 1px solid #aaa;
    background-color: #e5ac00;
    cursor: pointer; }
    .controller .controller-button:nth-child(2) {
      border-top: 1px solid #aaa; }
    .controller .controller-button:hover, .controller .controller-button:focus {
      outline: none;
      background-color: #efb400; }
    .controller .controller-button:active {
      font-size: 0.95em;
      background-color: #fabb00; }
    @media screen and (max-width: 770px) {
      .controller .controller-button {
        font-size: 0.9em; } }
  .controller .controller-pattern-dropdown {
    list-style-type: none;
    margin: 0;
    padding: 0; }
    .controller .controller-pattern-dropdown .controller-dropdown-element .controller-dropdown-button {
      display: block;
      width: 90%;
      margin: auto;
      color: #fafafa;
      border: none;
      padding: 0.3em;
      height: 1.8rem;
      font-size: 0.9em;
      font-family: "Montserrat", sans-serif;
      background-color: #efb400;
      border-bottom: 1px solid #aaa;
      cursor: pointer; }
      .controller .controller-pattern-dropdown .controller-dropdown-element .controller-dropdown-button:hover, .controller .controller-pattern-dropdown .controller-dropdown-element .controller-dropdown-button:focus {
        outline: none;
        background-color: #fabb00; }
      .controller .controller-pattern-dropdown .controller-dropdown-element .controller-dropdown-button:active {
        font-size: 0.85em;
        background-color: #ffc005; }
      @media screen and (max-width: 770px) {
        .controller .controller-pattern-dropdown .controller-dropdown-element .controller-dropdown-button {
          font-size: 0.8em; } }
