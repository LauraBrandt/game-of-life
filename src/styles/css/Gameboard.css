.gameboard {
  border: 3px solid rgba(0, 0, 0, 0.5);
  background-color: #fafafa;
  margin: 0 1rem; }

.board-row {
  margin: 0;
  padding: 0;
  font-size: 0;
  display: flex; }

.board-cell {
  height: 16px;
  width: 16px;
  display: inline-block;
  border-right: 1px solid #aaa;
  border-bottom: 1px solid #aaa; }
  @media screen and (max-width: 850px) {
    .board-cell {
      height: 14px;
      width: 14px; } }
  @media screen and (max-width: 750px) {
    .board-cell {
      height: 12px;
      width: 12px; } }
  @media screen and (max-width: 600px) {
    .board-cell {
      height: 9px;
      width: 9px; } }
  .board-cell.active {
    background-color: #222; }
