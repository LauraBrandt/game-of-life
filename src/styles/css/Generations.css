.generations {
  margin: 0 1em 0.5em 1em;
  font-size: 1.1em;
  color: #ffd65c;
  text-align: center;
  font-family: "Montserrat", sans-serif; }
  .generations span {
    color: #d7125e;
    border-radius: 3px;
    padding: 0.15em 0.55em;
    margin: 0.5em;
    font-family: 'Nova Mono', monospace;
    background-color: #ffc30f;
    font-size: 1.05em;
    text-shadow: 0 0 2px white; }
