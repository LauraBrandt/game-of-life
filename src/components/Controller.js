import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { tick, back, setRandom, setPattern, clearBoard } from '../actions/boardActions';
import { incrementGeneration, decrementGeneration, resetGeneration } from '../actions/generationActions';
import '../styles/css/Controller.css';
import patterns from '../startPatterns';

export class Controller extends Component {
  constructor() {
    super();

    this.state = {
      timer: null,
      isPlaying: false,
      dropdownOpen: false,
      patterns: []
    }

    this.togglePlay = this.togglePlay.bind(this);
    this.startTimer = this.startTimer.bind(this);
    this.clear = this.clear.bind(this);
    this.randomize = this.randomize.bind(this);
    this.setPattern = this.setPattern.bind(this);
    this.toggleDropdown = this.toggleDropdown.bind(this);
    this.goBack = this.goBack.bind(this);
    this.goForward = this.goForward.bind(this);
  }

  togglePlay() {
    if (this.state.isPlaying) {
      clearInterval(this.state.timer);
    } else {
      this.startTimer();
    }
    this.setState( prevState => ({ isPlaying: !prevState.isPlaying }) );
  }

  startTimer() {
    const timer = setInterval(() => {
      this.props.tick();
      this.props.incrementGeneration();
    }, 250);
    this.setState({ timer });
  }

  clear() {
    clearInterval(this.state.timer);
    this.setState({ isPlaying: false });
    this.props.clearBoard();
    this.props.resetGeneration();
  }

  randomize() {
    clearInterval(this.state.timer);
    this.setState({ isPlaying: false });
    this.props.resetGeneration();
    this.props.randomize();
  }

  setPattern(cells) {
    clearInterval(this.state.timer);
    this.setState({ isPlaying: false });
    this.props.resetGeneration();
    this.props.setPattern(cells);
  }

  toggleDropdown() {
    this.setState(prevState => ({dropdownOpen: !prevState.dropdownOpen}));
  }

  componentDidMount() {
    this.setState({ patterns });
    this.randomize();
  }

  componentWillUnmount() {
    clearInterval(this.state.timer);
  }

  goBack() {
    this.props.back();
    this.props.decrementGeneration();
  }

  goForward() {
    this.props.tick();
    this.props.incrementGeneration();
  }

  render() {
    return (
      <div className="controller">
        <div className="controller-controls">
          <p>Controls</p>
          <button className="controller-button" onClick={this.togglePlay}>
            {this.state.isPlaying ? 'Pause' : 'Play'}
          </button>
          <button className="controller-button" onClick={this.clear}>
            Clear
          </button>
          <div className="arrows">
            <button className="arrow arrow-left" onClick={this.goBack}>&#9664;</button>
            <p>Move 1 generation</p>
            <button className="arrow arrow-right" onClick={this.goForward}>&#9654;</button>
          </div>
        </div>
        
        <div className="controller-start-config">
          <p>Choose a starting configuration: <br/><span>(or click the squares on the board to make your own)</span></p>
          <button className="controller-button" onClick={this.randomize}>
            Random
          </button>
          <div className="controller-patterns">
            <button className="controller-button controller-pattern" onClick={this.toggleDropdown}>
              Patterns &#9662;
            </button>
            {this.state.dropdownOpen &&
              <ul className="controller-pattern-dropdown"> 
                {this.state.patterns.map(pattern => 
                  <li key={pattern.code} className="controller-dropdown-element">
                    <button className="controller-dropdown-button" onClick={this.setPattern.bind(null, pattern.cells)}>
                      {pattern.name}
                    </button>
                  </li>
                )}
              </ul>
            }
          </div>
        </div>
      </div>
    );
  }
}

Controller.propTypes = {
  tick: PropTypes.func.isRequired,
  back: PropTypes.func.isRequired,
  randomize: PropTypes.func.isRequired,
  setPattern: PropTypes.func.isRequired,
  clearBoard: PropTypes.func.isRequired,
  incrementGeneration: PropTypes.func.isRequired,
  decrementGeneration: PropTypes.func.isRequired,
  resetGeneration: PropTypes.func.isRequired
}

const mapDispatchToProps = dispatch => ({
  randomize: () => dispatch(setRandom()),
  setPattern: pattern => dispatch(setPattern(pattern)),
  tick: () => dispatch(tick()),
  back: () => dispatch(back()),
  clearBoard: () => dispatch(clearBoard()),
  incrementGeneration: () => dispatch(incrementGeneration()),
  decrementGeneration: () => dispatch(decrementGeneration()),
  resetGeneration: () => dispatch(resetGeneration()),
});

export default connect(null, mapDispatchToProps)(Controller);
