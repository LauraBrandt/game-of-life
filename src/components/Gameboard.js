import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { toggleCell } from '../actions/boardActions';
import '../styles/css/Gameboard.css';

export class Gameboard extends Component {
  render() {
    const { board, toggleCell } = this.props;

    return (
      <div className="gameboard">
        { board.map((row, i) => 
          <div className="board-row" key={i}>
            {row.map((cell, j) => 
              <div 
                className={`board-cell${cell ? ' active' : ''}`} 
                key={j} 
                onClick={toggleCell.bind(null, [i, j])}>
              </div>
            )}
          </div>
        )}
      </div>
    );
  }
}

Gameboard.propTypes = {
  board: PropTypes.array.isRequired,
  toggleCell: PropTypes.func
}

const mapStateToProps = state => ({
  board: state.board.current
});

const mapDispatchToProps = dispatch => ({
  toggleCell: cell => dispatch(toggleCell(cell))
});

export default connect(mapStateToProps, mapDispatchToProps)(Gameboard);
