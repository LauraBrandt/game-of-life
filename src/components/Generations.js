import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import '../styles/css/Generations.css';

export class Generations extends Component {
  render() {
    return (
      <div className="generations">
        Generation: <span>{this.props.generation}</span>
      </div>
    );
  }
}

Generations.propTypes = {
  generation: PropTypes.number.isRequired
}

const mapStateToProps = state => ({
  generation: state.generation
});

export default connect(mapStateToProps)(Generations);
