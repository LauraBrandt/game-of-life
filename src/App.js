import React, { Component } from 'react';
import Generations from './components/Generations';
import Gameboard from './components/Gameboard';
import Controller from './components/Controller';
import './styles/css/App.css';

class App extends Component {
  render() {
    return (
      <div className="app">
        <header className="app-header">
          <h1>Conway's Game of Life</h1>
        </header>
        <main className="app-main">
          <Generations />
          <Gameboard/>
          <Controller/>
        </main>
        <footer className="app-footer">
          <p>Designed and coded by <a href="https://github.com/LauraBrandt">Laura Brandt</a> as part of the <a href="https://learn.freecodecamp.org/coding-interview-prep/take-home-projects/build-the-game-of-life/">FreeCodeCamp</a> curriculum.</p>
        </footer>
      </div>
    );
  }
}

export default App;
