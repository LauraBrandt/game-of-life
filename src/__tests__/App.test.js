import React from 'react';
import {Provider} from 'react-redux';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import App from '../App';
import Generations from '../components/Generations'
import Gameboard from '../components/Gameboard';
import Controller from '../components/Controller';

describe('<App />', () => {
  describe('content', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<App />);
    });

    it('has a header', () => {
      expect(wrapper.find('header')).toHaveLength(1);
    });

    it('has a footer', () => {
      expect(wrapper.find('footer')).toHaveLength(1);
    });

    it('has 1 Generations component', () => {
      expect(wrapper.find(Generations)).toHaveLength(1);
    });

    it('has 1 Gameboard component', () => {
      expect(wrapper.find(Gameboard)).toHaveLength(1);
    });

    it('has 1 Controller component', () => {
      expect(wrapper.find(Controller)).toHaveLength(1);
    });
  });

  describe('snapshot', () => {
    it('renders correctly', () => {
      // given
      const mockStore = configureStore();
      const initialState = {
        generation: 0,
        board: {
          current: [],
          history: []
        }
      }
      const store = mockStore(initialState);
      // when
      const component = renderer.create(<Provider store={store}><App /></Provider>);
      const tree = component.toJSON();
      // then
      expect(tree).toMatchSnapshot();
    });
  });
});
