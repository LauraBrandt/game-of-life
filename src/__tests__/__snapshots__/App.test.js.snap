// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> snapshot renders correctly 1`] = `
<div
  className="app"
>
  <header
    className="app-header"
  >
    <h1>
      Conway's Game of Life
    </h1>
  </header>
  <main
    className="app-main"
  >
    <div
      className="generations"
    >
      Generation: 
      <span>
        0
      </span>
    </div>
    <div
      className="gameboard"
    />
    <div
      className="controller"
    >
      <div
        className="controller-controls"
      >
        <p>
          Controls
        </p>
        <button
          className="controller-button"
          onClick={[Function]}
        >
          Play
        </button>
        <button
          className="controller-button"
          onClick={[Function]}
        >
          Clear
        </button>
        <div
          className="arrows"
        >
          <button
            className="arrow arrow-left"
            onClick={[Function]}
          >
            ◀
          </button>
          <p>
            Move 1 generation
          </p>
          <button
            className="arrow arrow-right"
            onClick={[Function]}
          >
            ▶
          </button>
        </div>
      </div>
      <div
        className="controller-start-config"
      >
        <p>
          Choose a starting configuration: 
          <br />
          <span>
            (or click the squares on the board to make your own)
          </span>
        </p>
        <button
          className="controller-button"
          onClick={[Function]}
        >
          Random
        </button>
        <div
          className="controller-patterns"
        >
          <button
            className="controller-button controller-pattern"
            onClick={[Function]}
          >
            Patterns ▾
          </button>
        </div>
      </div>
    </div>
  </main>
  <footer
    className="app-footer"
  >
    <p>
      Designed and coded by 
      <a
        href="https://github.com/LauraBrandt"
      >
        Laura Brandt
      </a>
       as part of the 
      <a
        href="https://learn.freecodecamp.org/coding-interview-prep/take-home-projects/build-the-game-of-life/"
      >
        FreeCodeCamp
      </a>
       curriculum.
    </p>
  </footer>
</div>
`;
